### PR TITLE
Implement generic OIDC login

### PR DESCRIPTION
**Pull request by patrickmann**
This is the companion PR to https://github.com/Graylog2/graylog-plugin-enterprise/pull/2410 and https://github.com/Graylog2/graylog-plugin-cloud/pull/944

We introduce more generic naming in support of increased inheritance and code re-use.

/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#2410

### CODE REVIEW

**Comment by thll** [comment]
@patrickmann the diff appears to show many unrelated changes. could you rebase the PR to the current `master` please?

**Comment by thll** [comment]
From the description: 

> /jenkins-pr-deps Graylog2/graylog-plugin-enterprise#2410

I think we shouldn't have this dependency here. The build should pass without it, otherwise we'd have a problem for the non-enterprise version. Can we remove this?

**Comment by thll** [comment]
> From the description:
> 
> > /jenkins-pr-deps [Graylog2/graylog-plugin-enterprise#2410](https://github.com/Graylog2/graylog-plugin-enterprise/pull/2410)
> 
> I think we shouldn't have this dependency here. The build should pass without it, otherwise we'd have a problem for the non-enterprise version. Can we remove this?

My bad, sorry for the noise. I was confusing this with the github actions build step which passes just fine. I added the Jenkins statement again.

File: graylog2-server/src/main/java/org/graylog/security/authservice/AuthServiceBackend.java [hunk 1]
```diff
@@ -41,6 +41,19 @@
...
+    /**
+     * Authenticate user and provision a local user record
+     * @param token
+     * @param provisionerService
+     * @param isExternal user is managed by external service (e.g. Okta) and cannot be modified locally
+     * @return authentication details; empty if a problem was encountered
+     */
+    default Optional<AuthenticationDetails> authenticateAndProvision(AuthServiceToken token,
+                                                                     ProvisionerService provisionerService,
+                                                                     boolean isExternal) {
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog/security/authservice/AuthServiceBackend.java:56]: Do we still need this change? I don't see it being used with the current state of the implementation so I vote for reverting it to keep the interface clean.

File: graylog2-server/src/main/java/org/graylog/security/authservice/AuthenticationDetails.java [hunk 2]
```diff
@@ -32,6 +32,8 @@ public static Builder builder() {
         return Builder.create();
     }
 
+    public abstract AuthenticationDetails.Builder toBuilder();
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog/security/authservice/AuthenticationDetails.java:35]: It's nice to have this, but we don't have code calling this. If we decide to keep the PR for the UI changes, we should leave this change in.

**Review by thll — CHANGES_REQUESTED** [review]
As the backend code depending on this PR has been refactored, I think from a backend perspective this PR is now obsolete. We could keep the change in `AuthenticationDetails.java` though.

I didn't look at the UI changes.

**Review by thll — APPROVED** [review]
Backend changes are fine.
Feel free to merge once the frontend changes have been reviewed.

File: graylog2-web-interface/src/logic/authentication/okta/types.ts [hunk 3]
```diff
@@ -20,22 +20,25 @@ import * as Immutable from 'immutable';
 import AuthenticationBackend from 'logic/authentication/AuthenticationBackend';
 
 export interface OktaBackendConfig {
-  type: 'okta';
-  oktaBaseUrl: string;
+  type: string;
+  baseUrl?: string;
+  oktaBaseUrl?: string;
```
Comment (edmundoa) [inline graylog2-web-interface/src/logic/authentication/okta/types.ts:25]: Should we change this types like we did in enterprise? That is, to have one for OIDC and one for Okta.

**Review by edmundoa — APPROVED** [review]
LGTM, thank you for all the work! 👍

**Pull request merged**